File: analytics/src/main/java/com/bluetriangle/analytics/PayloadCache.java
package com.bluetriangle.analytics;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

final class PayloadCache {
    private final BlueTriangleConfiguration configuration;
    private final @NonNull File directory;
    private final int maxSize;

    public PayloadCache(BlueTriangleConfiguration configuration) {
        this.configuration = configuration;
        directory = new File(configuration.getCacheDirectory());
        if (isDirectoryValid()) {
            maxSize = configuration.getMaxCacheItems();
        } else {
            maxSize = 0;
        }
    }

    /**
     * Get the next cached payload to attempt resending, if any
     *
     * @return next payload to send or null if none
     */
    @Nullable
    public synchronized Payload getNextCachedPayload() {
        final File[] files = getAllCacheFiles(true);
        if (files.length > 0) {
            final File file = files[0];
            final Payload payload = readPayload(file);
            if(!file.delete()) {
                configuration.getLogger().error("error deleting next cached payload file %s", file.getName());
            }
            return payload;
        }
        return null;
    }

    /**
     * Check to see if there are any cached payloads that need to still be sent
     *
     * @return true if there are existing payload cache files, else false
     */
    public boolean hasCachedPayloads() {
        if (maxSize > 0) {
            return getAllCacheFiles(false).length > 0;
        }
        return false;
    }

    /**
     * Clears the cache by deleting all cache files in the cache directory
     */
    public void clearCache() {
        for (final File file : getAllCacheFiles(false)) {
            if (!file.delete()) {
                configuration.getLogger().error("Error deleting %s while clearing cache", file.getName());
            }
        }
    }

    public void cachePayload(@NonNull final Payload payload) {
        if (maxSize > 0) {
            if (payload.payloadAttempts >= configuration.getMaxAttempts()) {
                configuration.getLogger().warn("Payload %s has exceeded max attempts %s", payload.id, payload.payloadAttempts);
                return;
            }
            try {
                payload.serialize(directory);
            } catch (IOException e) {
                configuration.getLogger().error(e, "Failed to cache payload %s", payload.id);
            }
            rotateCacheIfNeeded();
        }
    }

    @Nullable
    public Payload readPayload(@NonNull final File file) {
        try {
            return Payload.deserialize(file);
        } catch (IOException e) {
            configuration.getLogger().error(e, "Failed to load payload %s", file.getAbsolutePath());
            if (!file.delete()) {
                configuration.getLogger().warn("Could not delete payload file %s", file.getAbsolutePath());
            }
        }
        return null;
    }

    /**
     * Get all cache files, sorted oldest to newest
     *
     * @return all cached
     */
    @NonNull
    private File[] getAllCacheFiles(final boolean sort) {
        if (isDirectoryValid()) {
            // lets filter the session.json here
            final File[] files = directory.listFiles();
            if (files != null) {
                if (sort) {
                    sortFilesOldestToNewest(files);
                }
                return files;
            }
        }
        return new File[]{};
    }

    /**
     * Check if a dir. is valid and have write and read permission
     *
     * @return true if valid and has permissions or false otherwise
     */
    private boolean isDirectoryValid() {
        if (!directory.isDirectory() || !directory.canWrite() || !directory.canRead()) {
            configuration.getLogger().error("The directory for caching files is not valid: %s", directory.getAbsolutePath());
            return false;
        }
        return true;
    }

    /**
     * Sort files from oldest to the newest using the lastModified method
     *
     * @param files the Files
     */
    private void sortFilesOldestToNewest(@NonNull File[] files) {
        // just sort it if more than 1 file
        if (files.length > 1) {
            Arrays.sort(files, (f1, f2) -> Long.compare(f1.lastModified(), f2.lastModified()));
        }
    }

    /**
     * Rotates the caching folder if full, deleting the oldest files first
     */
    void rotateCacheIfNeeded() {
        final File[] files = getAllCacheFiles(true);
        final int length = files.length;
        if (length >= maxSize) {
            configuration.getLogger().warn("Cache folder size %s > %s. Rotating files.", length, maxSize);
            final int totalToBeDeleted = (length - maxSize) + 1;

            sortFilesOldestToNewest(files);

            // delete files from the top of the Array as its sorted by the oldest to the newest
            for (int i = 0; i < totalToBeDeleted; i++) {
                final File file = files[i];
                if (!file.delete()) {
                    configuration.getLogger().warn("Error deleting file: %s", file.getAbsolutePath());
                }
            }
        }
    }
}

File: analytics/src/main/java/com/bluetriangle/analytics/PayloadCache.kt
package com.bluetriangle.analytics

import com.bluetriangle.analytics.caching.provider.PayloadProviderImpl
import com.bluetriangle.analytics.caching.classifier.ExtensionClassifier
import com.bluetriangle.analytics.caching.CacheLimitState
import com.bluetriangle.analytics.caching.MemoryLimitVerifier
import com.bluetriangle.analytics.caching.classifier.CacheType
import com.bluetriangle.analytics.caching.classifier.Classifier
import com.bluetriangle.analytics.caching.provider.PayloadComponent
import com.bluetriangle.analytics.caching.provider.PayloadProvider
import com.bluetriangle.analytics.caching.serializer.PayloadSerializer
import com.bluetriangle.analytics.caching.serializer.Serializer
import com.bluetriangle.analytics.utility.isDirectoryInvalid
import com.bluetriangle.analytics.utility.mb
import java.io.File
import java.io.IOException

class PayloadCache(private val configuration: BlueTriangleConfiguration) {

    companion object {
        val pickOrder = listOf(CacheType.Analytics, CacheType.Error, CacheType.Other, CacheType.Wcd)
        val deleteOrder =
            listOf(CacheType.Wcd, CacheType.Other, CacheType.Analytics, CacheType.Error)
    }

    private var cacheExpiryDuration = configuration.cacheExpiryDuration

    private var cacheMemoryLimit = configuration.cacheMemoryLimit

    private val directory: File = File(configuration.cacheDirectory!!)
    private var maxSize = if (isDirectoryValid) {
        configuration.maxCacheItems
    } else {
        0
    }
    private val memoryLimitVerifier =
        MemoryLimitVerifier(configuration.logger, cacheMemoryLimit, directory)
    private val classifier: Classifier =
        ExtensionClassifier()
    private val serializer: Serializer =
        PayloadSerializer(configuration.logger, directory, classifier)
    private val cachePayloadProvider: PayloadProvider =
        PayloadProviderImpl(directory, serializer, cacheExpiryDuration)

    /**
     * Get the next cached payload to attempt resending, if any
     *
     * @return next payload to send or null if none
     */
    fun pickNext(): Payload? {
        clearExpired()
        val file = getNextFile(pickOrder) ?: return null
        val payload = readPayload(file)
        configuration.logger?.debug("Next cached payload file: ${file.name}, ${payload?.type}")
        if (!file.delete()) {
            configuration.logger?.error("error deleting next cached payload file ${file.name}")
        }
        return payload
    }

    private fun getNextFile(order: List<CacheType>): File? {
        var component: PayloadComponent? = null
        for (type in order) {
            component = cachePayloadProvider.getByType(type).minByOrNull { it.createdAt }
            if (component != null) {
                break
            }
        }
        return component?.file
    }

    /**
     * Check to see if there are any cached payloads that need to still be sent
     *
     * @return true if there are existing payload cache files, else false
     */
    fun hasCachedPayloads(): Boolean {
        return getAllCacheFiles(false).isNotEmpty()
    }

    /**
     * Clears the cache by deleting all cache files in the cache directory
     */
    fun clearCache() {
        for (file in getAllCacheFiles(false)) {
            if (!file.delete()) {
                configuration.logger?.error("Error deleting ${file.name} while clearing cache")
            }
        }
    }

    fun save(payload: Payload) {
        try {
            serializer.serialize(payload)
        } catch (e: IOException) {
            configuration.logger?.error(e, "Failed to cache payload ${payload.id}")
        }
        cleanUp()
    }

    private fun readPayload(file: File): Payload? {
        try {
            return serializer.deserialize(file)
        } catch (e: IOException) {
            configuration.logger?.error(e, "Failed to load payload ${file.absolutePath}")
            if (!file.delete()) {
                configuration.logger?.warn("Could not delete payload file ${file.absolutePath}")
            }
        }
        return null
    }

    /**
     * Get all cache files, sorted oldest to newest
     *
     * @return all cached
     */
    private fun getAllCacheFiles(sort: Boolean): List<File> {
        if (isDirectoryValid) {
            // lets filter the session.json here
            val files = directory.listFiles()?.toList() ?: emptyList()
            if (sort) {
                return files.sortedBy { it.lastModified() }
            }
            return files
        }
        return emptyList()
    }

    /**
     * Check if a dir. is valid and have write and read permission
     *
     * @return true if valid and has permissions or false otherwise
     */
    private val isDirectoryValid: Boolean
        get() {
            if (directory.isDirectoryInvalid) {
                configuration.logger?.error("The directory for caching files is not valid: ${directory.absolutePath}")
                return false
            }
            return true
        }

    /**
     * Rotates the caching folder if full, deleting the oldest files first
     */
    private fun cleanUp() {
        clearExpired()
        var limitReached = memoryLimitVerifier.isLimitReached()
        while (limitReached is CacheLimitState.Reached) {
            configuration.logger?.warn("Cache memory limit exceeded by ${limitReached.exceededBytes.mb}MB! Deleting files")
            val fileToDelete = getNextFile(deleteOrder)
            if (fileToDelete == null || !fileToDelete.delete()) break
            limitReached = memoryLimitVerifier.isLimitReached()
        }
    }

    private fun clearExpired() {
        cachePayloadProvider.getExpired().forEach {
            it.file.delete()
        }
    }
}